translator/util/ec2util/ec2util.go

```diff
@@ -67,46 +71,49 @@ func initEC2UtilSingleton() (newInstance *ec2Util) {
 		log.Println("W! [EC2] Sleep until network is up")
 		time.Sleep(1 * time.Second)
 	}
+
 	if !networkUp {
 		log.Println("E! [EC2] No available network interface")
 	}
 
-	ses, err := session.NewSession()
+	err := newInstance.deriveEC2MetadataFromIMDS()
+
 	if err != nil {
-		log.Println("E! [EC2] getting new session info: ", err)
-		return
+		log.Println("E! [EC2] Cannot get EC2 Metadata from IMDS:", err)
 	}
-	md := ec2metadata.New(ses)
 
-	if !md.Available() {
-		log.Println("E! ec2metadata is not available")
-		return
-	}
+	return
+}
 
-	if info, err := md.GetMetadata("instance-id"); err == nil {
-		newInstance.InstanceID = info
-	} else {
-		log.Println("E! getting instance-id from EC2 metadata fail: ", err)
+func (e *ec2Util) deriveEC2MetadataFromIMDS() error {
+	ses, err := session.NewSession()
+
+	if err != nil {
+		return err
 	}
 
-	if info, err := md.GetMetadata("hostname"); err == nil {
-		newInstance.Hostname = info
-	} else {
-		log.Println("E! getting hostname from EC2 metadata fail: ", err)
+	md := ec2metadata.New(ses)
+
+	if !md.Available() {
+		return errors.New("EC2 metadata is not available.")
 	}
 
-	if info, err := md.GetMetadata("local-ipv4"); err == nil {
-		newInstance.PrivateIP = info
+	// More information on API: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html#instance-metadata-ex-2
+	if hostname, err := md.GetMetadata("hostname"); err == nil {
+		e.Hostname = hostname
 	} else {
-		log.Println("E! getting local-ipv4 from EC2 metadata fail: ", err)
+		log.Println("E! [EC2] Fetch hostname from EC2 metadata fail:", err)
 	}
 
-	if info, err := md.GetInstanceIdentityDocument(); err == nil {
-		newInstance.Region = info.Region
-		newInstance.AccountID = info.AccountID
+	// More information on API: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
+	if instanceIdentityDocument, err := md.GetInstanceIdentityDocument(); err == nil {
+		e.Region = instanceIdentityDocument.Region
+		e.AccountID = instanceIdentityDocument.AccountID
+		e.PrivateIP = instanceIdentityDocument.PrivateIP
+		e.InstanceID = instanceIdentityDocument.InstanceID
```

Comment:
This is a good catch to consolidate the calls to IMDS to help avoid potential throttling. 

Have you confirmed in testing that using the identity document to get the local ipv4 and instance id show no regression/change compared to the existing way of getting that info?

Reply:
Yes. There are no regression change between these two.
* **Before the changes:**
```
****** processing amazon-cloudwatch-agent ******
/opt/aws/amazon-cloudwatch-agent/bin/config-downloader --output-dir /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d --download-source file:./config.json --mode ec2 --config /opt/aws/amazon-cloudwatch-agent/etc/common-config.toml --multi-config default
Got Home directory: /root
I! Set home dir Linux: /root
I! SDKRegionWithCredsMap region:  us-west-2
Successfully fetched the config and saved in /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d/file_config.json.tmp
Start configuration validation...
/opt/aws/amazon-cloudwatch-agent/bin/config-translator --input /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json --input-dir /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d --output /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.toml --mode ec2 --config /opt/aws/amazon-cloudwatch-agent/etc/common-config.toml --multi-config default
2022/06/10 03:23:20 Reading json config file path: /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d/file_config.json.tmp ...
2022/06/10 03:23:20 I! Valid Json input schema.
I! Detecting run_as_user...
Got Home directory: /root
I! Set home dir Linux: /root
I! SDKRegionWithCredsMap region:  us-west-2
2022/06/10 03:23:20 D! [EC2] Found active network interface
2022/06/10 03:23:20 Instance ID i-005113ad6a00082e0
2022/06/10 03:23:20 Hostname ip-10-189-44-37.amazon.com
2022/06/10 03:23:20 Private Ip 10.189.44.37
```
* **After the changes**
```
****** processing amazon-cloudwatch-agent ******
/opt/aws/amazon-cloudwatch-agent/bin/config-downloader --output-dir /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d --download-source file:./config.json --mode ec2 --config /opt/aws/amazon-cloudwatch-agent/etc/common-config.toml --multi-config default
Got Home directory: /root
I! Set home dir Linux: /root
I! SDKRegionWithCredsMap region:  us-west-2
Successfully fetched the config and saved in /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d/file_config.json.tmp
Start configuration validation...
/opt/aws/amazon-cloudwatch-agent/bin/config-translator --input /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json --input-dir /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d --output /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.toml --mode ec2 --config /opt/aws/amazon-cloudwatch-agent/etc/common-config.toml --multi-config default
2022/06/10 03:56:00 Reading json config file path: /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d/file_config.json.tmp ...
2022/06/10 03:56:00 I! Valid Json input schema.
I! Detecting run_as_user...
Got Home directory: /root
I! Set home dir Linux: /root
I! SDKRegionWithCredsMap region:  us-west-2
2022/06/10 03:56:00 D! [EC2] Found active network interface
2022/06/10 03:56:00 Private Ip nay 10.189.44.37
2022/06/10 03:56:00 Instance ID nay i-005113ad6a00082e0
No csm configuration found.
Configuration validation first phase succeeded
/opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent -schematest -config /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.toml
Configuration validation second phase succeeded
Configuration validation succeeded
Redirecting to /bin/systemctl stop amazon-cloudwatch-agent.service
Redirecting to /bin/systemctl restart amazon-cloudwatch-agent.service
```